File: task-generator-julia/task-2.jl
using Random
using LinearAlgebra

# Функция для генерации матрицы и вектора
function generate_system(A_size, range)
    A = rand(range, A_size...)  # Генерация матрицы коэффициентов 4x4
    b = rand(range, A_size[1])  
    return A, b
end

# Функция для применения формулы Крамера для x3
function solve_for_x(A, b, x_n)
    A3 = copy(A)            # Копируем A для изменения
    A3[:, x_n] = b           # Заменяем третий столбец на вектор b
    det_A = round(det(A),digits = 5)  # Главный определитель
    det_An = round(det(A3),digits = 5) # Определитель для x3
    x = "$det_An/$det_A "   # Значение x3 по формуле Крамера
    return det_A, det_An, x
end

function makesystem(A,b)
    m, n = size(A)
    eq = "{█("
    for i in 1:m
        str = "&"
        first_elem = false
        for j in 1:n
            if first_elem * (A[i,j] > 0)
                str *= "+" 
            elseif A[i,j] == -1
                str *= "-"
            end
            if A[i,j] != 0
                first_elem = true
                if abs(A[i,j]) != 1
                    str *= "$(A[i,j])"
                end
                str *= "x_$j"
            end
        end
        eq *= str * "=$(b[i])"
        if i < m 
            eq *= "@"
        end
    end
    return eq * ")┤"
end

# open("task2_output.txt", "w") do file
#     write(file,"Задание 2\n\n")
# end

# for i in 1:5
    
#     A, b = nothing, nothing
    
#     # Генерация системы до получения невырожденной
#     while true
#         A, b = generate_system()
#         if det(A) != 0  # Проверка на невырожденность
#             break
#         end
#     end

#     # Нахождение определителей и значения x3
#     det_A, det_A3, x3 = solve_for_x3(A, b)
#     system = makesystem(A,b)

#     open("task2_output.txt", "a") do file
#         write(file,"Вариант $i\n")
#         write(file,system,"\n")
#         write(file,"Δ = $det_A\nΔ₃ = $det_A3\nx₃ = $x3","\n\n")
#     end

#     # display(A)
#     # display(b)
#     println(system)
    
#     println("Δ = $det_A\nΔ₃ = $det_A3\nx₃ = $x3")
# end

function gen_task_2(A_size, range, x_n)
    A, b = nothing, nothing
    while true
        A, b = generate_system(A_size, range)
        if det(A) != 0  # Проверка на невырожденность
            break
        end
    end
    
    # Нахождение определителей и значения x3
    det_A, det_A3, x3 = solve_for_x(A, b, x_n)
    system = makesystem(A,b)
#     println(system)    
#     println("Δ = $det_A\nΔ₃ = $det_A3\nx₃ = $x3")
    answer = "Δ = $det_A\nΔ₃ = $det_A3\nx₃ = $x3"
    return (system, answer)
end

File: task-generator/task2.py
import numpy as np

# Function to generate matrix and vector
def generate_system(A_size, range):
    A = np.random.randint(range[0], range[1], A_size)  # Generate coefficient matrix
    b = np.random.randint(range[0], range[1], A_size[0])  
    return A, b

# Function to apply Cramer's rule for x3
def solve_for_x(A, b, x_n):
    A3 = A.copy()            # Copy A for modification
    A3[:, x_n] = b           # Replace the third column with vector b
    det_A = round(np.linalg.det(A), 5)  # Main determinant
    det_An = round(np.linalg.det(A3), 5) # Determinant for x3
    x = f"{det_An}/{det_A}"   # Value of x3 by Cramer's rule
    return det_A, det_An, x

# def makesystem(A, b):
#     m, n = A.shape
#     eq = "{█("
#     for i in range(m):
#         str_ = "&"
#         first_elem = False
#         for j in range(n):
#             if first_elem and A[i, j] > 0:
#                 str_ += "+" 
#             elif A[i, j] == -1:
#                 str_ += "-"
#             if A[i, j] != 0:
#                 first_elem = True
#                 if abs(A[i, j]) != 1:
#                     str_ += f"{A[i, j]}"
#                 str_ += f"x_{j + 1}"
#         eq += str_ + f"={b[i]}"
#         if i < m - 1:
#             eq += "@"
#     return eq + ")┤"

# Make system of equations
def makesystem(A, b):
    m, n = A.shape
    eq = " "
    for i in range(m):
        str_ = " "
        first_elem = False
        for j in range(n):
            if first_elem and A[i, j] > 0:
                str_ += " + " 
            elif A[i, j] == -1:
                str_ += " - "
            if A[i, j] != 0:
                first_elem = True
                if abs(A[i, j]) != 1:
                    str_ += f"{A[i, j]}"
                str_ += f"x_{j + 1}"
        eq += str_ + f" = {b[i]}\n"
    return eq

# Make system and solution
def gen_task_2(A_size, range, x_n):
    A, b = None, None
    while True:
        A, b = generate_system(A_size, range)
        if np.linalg.det(A) != 0:  # Check for non-degeneracy
            break
    
    # Finding determinants and value of x3
    det_A, det_A3, x3 = solve_for_x(A, b, x_n)
    system = makesystem(A, b)
    answer = f"\nΔ = {det_A}\nΔ₃ = {det_A3}\nx₃ = {x3}"
    return system, answer

